moved config to http-vue-loader

OmgImAlexis opened the pull request:
@p0psicles

sharkykh commented:
IMO, we have to do something about the fact it loads the config from the API twice.
![image](https://user-images.githubusercontent.com/10238474/41467998-2658d34c-70b1-11e8-906e-43013acd915e.png)
One from `core.js` and the other from the store.

OmgImAlexis commented on themes-default/slim/static/js/templates/config.vue:
```diff
@@ -0,0 +1,58 @@
+    computed: Object.assign(store.mapState(['config']), {
+    }),
+    mounted() {
+        store.dispatch('getConfig');
```
This isn't needed if the parent Vue instance already has access to the config which it would once I move it. I just figured I'd start with moving the config page as is.

OmgImAlexis commented on themes-default/slim/views/config.mako:
```diff
@@ -1,110 +1,12 @@
-        };
-    },
-    async mounted() {
-        const { data } = await api.get('config/main');
```
If you notice this would already hit the config twice since it's got the api check here. I just moved it to a store dispatch instead of a raw api method.

OmgImAlexis commented:
Okay, this doesn't work if the component uses another component. I'm not sure if there's anyway to fix that so maybe we should skip this for now?

OmgImAlexis commented:
I've added the 404 route and added it to the global VueRouter. That's also been moved to it's own file.

---

Once a `.mako` file has been moved over completely to Vue the template can be switched to `index.mako` as it's setup to load as a SPA.

---

`Vue-meta` should be removed once all the pages are added to the router but it's needed still otherwise the title of each page doesn't get set.

labrys commented:
@pymedusa/support I don't know if anyone's addressed it but with so many things moving to javascript you should implement an issue submitter and in app logs for all the js code.

p0psicles commented:
As soon as we can ditch the mako files. We can start using webpack. And with that unit and end2end test our vue components.

Were not that far off. With allot of pahlges already transitioned to vue.

sharkykh commented:
I already have a local branch that automatically logs Vue errors to Python.
I just wasn't sure about my code.

sharkykh commented on themes-default/slim/static/js/templates/config.vue:
```diff
@@ -0,0 +1,55 @@
+<template>
+    <div id="config-content">
+        <table class="infoTable" cellspacing="1" border="0" cellpadding="0" width="100%">
+            <tr v-if="config.release">
```
My changes from #4477 were reverted here.

sharkykh commented on themes-default/slim/static/js/lib/http-vue-loader.min.js:
```diff
@@ -0,0 +1,470 @@
+(function umd(root,factory){
```
This file is named `.min.js` yet it is not minimised.

sharkykh commented on medusa/server/web/config/handler.py:
```diff
@@ -46,7 +46,7 @@ def index(self):
-        t = PageTemplate(rh=self, filename='config.mako')
+        t = PageTemplate(rh=self, filename='index.mako')
 
         try:
```
This route can be simplified now.
Many of the values are already served from the API.

The route function can be:
```py
    def index(self):
        """
        Render the Help & Info page.

        [Converted to VueRouter]
        """
        t = PageTemplate(rh=self, filename='index.mako')
        return t.render(submenu=self.ConfigMenu(), topmenu='config')
```

sharkykh replied:
Oh yeah, and also remove the unused imports at the top of the file.

sharkykh commented on themes-default/slim/static/js/templates/config.vue:
```diff
@@ -0,0 +1,55 @@
+<script>
+module.exports = {
+    name: 'config',
+    computed: Object.assign(store.mapState(['config']), {
```
I don't see why this and the `dispatch` below are needed here.
Why did you add the following mixin if not for that?
https://github.com/pymedusa/Medusa/blob/f535c3d89c28e97861ec5cf6cb36f8d725d21fb2/themes-default/slim/views/layouts/main.mako#L163-L177

sharkykh commented on medusa/server/web/config/handler.py:
```diff
@@ -2,16 +2,7 @@
 
 """Base handler for Config pages."""
 
-from __future__ import unicode_literals
```
`__future__` imports should usually stay. This one in particular.
Basically, they're the first steps of making Python 2 compatible with Python 3.
Sorry for not being more explicit...

sharkykh commented on themes-default/slim/static/js/templates/config.vue:
```diff
@@ -0,0 +1,68 @@
+    computed: store.mapState(['config']),
+    mounted() {
+        const { $store } = this;
+        $store.dispatch('getConfig');
```
You didn't respond to my review about that.
I don't understand why we still need to get the config if we enabled it globally with the mixin?

OmgImAlexis replied:
I have to look into it but for some reason it's not getting the config set to anything but `null` like it is when we first load the store.

sharkykh replied:
I removed that code and it still works fine for me? :thinking: 
![2018-07-02_10-10-59](https://user-images.githubusercontent.com/10238474/42149796-44dd4460-7de0-11e8-9b02-47424688bddd.gif)

OmgImAlexis replied:
Who knows then.. shall I remove it?

OmgImAlexis commented:
After trying to convert more templates locally I found that the jQuery tabs refuse to load if a Vue HTML element is found inside of it and if I move the component back to being loaded locally via the x-templates it works fine.

sharkykh commented:
@OmgImAlexis 
Have you tried calling the tabs initializations (`$(element).tabs()` IIRC) on the component `mounted()`?

OmgImAlexis commented:
Yes, I also tried commenting it out and running it after everything has loaded via the console and nothing happens.

sharkykh commented on themes-default/slim/static/js/templates/config.vue:
```diff
@@ -0,0 +1,64 @@
+
+module.exports = {
+    name: 'config',
+    computed: store.mapState(['config']),
```
I don't know if it affects performance, but this mapState is not needed.

OmgImAlexis replied:
I still want to keep it there as I'd like to disable the mixin for `.vue` files.

sharkykh commented on themes-default/slim/static/js/templates/config.vue:
```diff
@@ -0,0 +1,64 @@
+<template>
+    <div id="config-content">
```
How do you feel about:
```vue
    <div v-if="!config.pythonVersion" class="vue-loading" id="config-content">
        <h3>Loading...</h3>
    </div>
    <div v-else id="config-content">
```

It gives it a cleaner look in the meantime, but a little more work when it's no longer needed.
I added `class="vue-loading"` so if we keep it the same on all `.vue` files it would be easy to remove them.

OmgImAlexis replied:
I was thinking we could just add a loading component before the `<router-view>`?

```
<loading v-if="loading"></loading>
<router-view v-else></router-view>

sharkykh replied:
How would you know when the values were loaded?

OmgImAlexis replied:
Well for example on this you'd set `loading: true` on data and then set a function to watch `config` when it updates set `loading: false`.

sharkykh commented on themes-default/slim/static/js/lib/http-vue-loader.js:
```diff
@@ -0,0 +1,470 @@
+(function umd(root,factory){
```
I still prefer a minimised version of this file, unless you believe it would be easier to debug issues?

OmgImAlexis replied:
Yeah, I think for now it'll be easier.

sharkykh commented on themes/dark/templates/config.mako:
```diff
@@ -1,122 +1,5 @@
 <%inherit file="/layouts/main.mako"/>
```
These files didn't get deleted on dark and light themes, need to manually remove

OmgImAlexis commented on themes-default/slim/static/js/router.js:
```diff
@@ -0,0 +1,47 @@
+            component: notFoundComponent
+        }, {
+            path: '*',
+            redirect: '/not-found'
```
Not getting this issue on dev with and without webroot so I'm not sure what's causing it on your end.Do you maybe still have the service worker loaded?

sharkykh replied:
I don't use a webroot. ServiceWorker is loaded but I don't it has anything to do with it.

Try navigating from say, `/home` to `/schedule` by clicking the Schedule nav link. You should get this issue.
Don't know why it only seems to happen on `/schedule` route.

OmgImAlexis replied:
That 100% has something todo with it. Remove the service worker.

sharkykh replied:
Oh, I just figured it out.
https://github.com/pymedusa/Medusa/blob/9eaf64841e85608af4ea1cff3a516237a9ef7bed/themes/light/templates/schedule.mako#L15
We need to remove this line.

OmgImAlexis replied:
Oh.. no we just need to add the route for it. I'm moving the headers and titles over to the `router.js`. Just don't add a component property and it works fine.

For example.

```js
routes: Object.assign(routes || [], [{
    path: '/config',
    name: 'config',
    meta: {
        title: 'Help & Info',
        header: 'Medusa Configuration'
    },
    component: configComponent
}, {
    path: '/home',
    name: 'home',
    meta: {
        title: 'Home',
        header: 'Show list'
    }
}, {
    path: '/history',
    name: 'history',
    meta: {
        title: 'History',
        header: 'History'
    }
}, {
    path: '/not-found',
    name: 'not-found',
    meta: {
        title: '404',
        header: '404 - page not found'
    },
    component: notFoundComponent
}, {
    path: '*',
    redirect: '/not-found'
}])
```

OmgImAlexis replied:
We then use this in the pages that aren't `.vue` since they don't use the global Vue.

```
header() {
    return this.$route.meta.header;
},
```

sharkykh replied:
This is the only page that has the `router` as a property now.
I still think removing it is the correct option since it'll be a while until schedule is moved to the router.
